meson.py

```diff
@@ -70,6 +70,11 @@ def build_ninja():
   # Set up the build environment, i.e. clone or download all submodules
   init_submodules('auto')
 
+  # Exit after setting up submodules, if system-wide Meson and Ninja are
```

Comment:
If you use the system wide meson how is this script called?

Reply:
> If you use the system wide meson how is this script called?

As documented in the commit message, a user may run ./meson.py with the variable SU2_MESON set to 'no' and then run system-wide Meson and Ninja.

Reply:
Thanks I did not expand the message.
This may not be very intuitive.
Wdyt of adding a preconfigure.py script? That way "our meson" can call it with an argument whereas calling it by itself will do what the env var is doing now?
If you know a way of including that preconfigure step in meson.build that may even be better

Reply:
You can have the second line in meson.build (after declaring the project() function) be a run_command() that executes your preconfigure program.

I think all that it actually does at the moment is:
- build ninja
- download submodules

You can assume if the user is using meson directly that the user has acquired some copy of ninja already. So all you need to do is download the submodules.

(Meson has built-in functionality to download submodules on demand, but only if those submodules are used via the subprojects/ directory.)

Reply:
> Thanks I did not expand the message. This may not be very intuitive. Wdyt of adding a preconfigure.py script? That way "our meson" can call it with an argument whereas calling it by itself will do what the env var is doing now?

I can try to implement this strategy. In case I succeed, I will open a separate pull request.

Reply:
Great, you can push to this PR it doesn't have to be separate.

Reply:
> Great, you can push to this PR it doesn't have to be separate.

It would be a different implementation, almost starting from scratch: that's why I thought I had to create a distinct topic branch and open a separate PR.

But, if you prefer that the whole history of this first attempt remains in the SU2 git repository, although superseded by one or more subsequent commits implementing the second attempt, I can push to the already existing topic branch.

Please let me know, in case you prefer that I just continue to accumulate commits in this same PR...

Reply:
Erm...

... why?

Why would the history of an *attempt* be obligated to remain in the PR? Just rebase, amend, and reorganize your work until the final result represents what should appear in the SU2 repository.

It's a defensible choice to create a brand new topic branch locally, and then force push it to the existing PR. :)

Reply:
We don't have such restrictions @frx-wintermute you can continue here.
You can always squash merge at the end if we want (but we do not require it).

Reply:
> We don't have such restrictions @frx-wintermute you can continue here. You can always squash merge at the end if you want (but we do not require it).

OK, thanks for the clarification @pcarruscag: I have just pushed to the already existing topic branch.
Please review the implementation.

I have tested the build and I obtain bit-for-bit identical executables with the default "self-contained" approach (./meson.py + ./ninja) in both the master branch and my topic branch, and with the system-wide Meson and Ninja approach (./preconfigure.py + meson + ninja) in my topic branch. Provided I pass the same configuration options to ./meson.py or to meson, of course.

See the last commit message for more details on how to use one or the other possible approaches.